README.md
[![Build
Status](https://travis-ci.org/trinker/cal.svg?branch=master)](https://travis-ci.org/trinker/cal)
[![Coverage
Status](https://coveralls.io/repos/trinker/cal/badge.svg?branch=master)](https://coveralls.io/r/trinker/cal?branch=master)

**cal** is an R package designed to partially mimic the [Unix
`cal`](https://www.tutorialspoint.com/unix_commands/cal.htm) command
line tool for generating ASCII calendars in the console.

Installation
============

To download the development version of **cal**:

Download the [zip ball](https://github.com/trinker/cal/zipball/master)
or [tar ball](https://github.com/trinker/cal/tarball/master), decompress
and run `R CMD INSTALL` on it, or use the **pacman** package to install
the development version:

    if (!require("pacman")) install.packages("pacman")
    pacman::p_load_current_gh("trinker/cal")

Examples:
=========

    cal()

                                    2018                               

            March                  April                   May                               
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa                       
                 1  2  3    1  2  3  4  5  6  7          1  2  3  4  5                       
     4  5  6  7  8  9 10    8  9 10 11 12 13 14    6  7  8  9 10 11 12                       
    11 12 13 14 15 16 17   15 16 17 18 19 20 21   13 14 15 16 17 18 19                       
    18 19 20 21 22 23 24   22 23 24 25 26 27 28   20 21 22 23 24 25 26                       
    25 26 27 28 29 30 31   29 30                  27 28 29 30 31                             
                                                                                             

    cal("2017")

                                               2017                                           

           January               February                 March                  April       
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
     1  2  3  4  5  6  7             1  2  3  4             1  2  3  4                      1
     8  9 10 11 12 13 14    5  6  7  8  9 10 11    5  6  7  8  9 10 11    2  3  4  5  6  7  8
    15 16 17 18 19 20 21   12 13 14 15 16 17 18   12 13 14 15 16 17 18    9 10 11 12 13 14 15
    22 23 24 25 26 27 28   19 20 21 22 23 24 25   19 20 21 22 23 24 25   16 17 18 19 20 21 22
    29 30 31               26 27 28               26 27 28 29 30 31      23 24 25 26 27 28 29
                                                                         30                  
             May                   June                   July                  August       
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
        1  2  3  4  5  6                1  2  3                      1          1  2  3  4  5
     7  8  9 10 11 12 13    4  5  6  7  8  9 10    2  3  4  5  6  7  8    6  7  8  9 10 11 12
    14 15 16 17 18 19 20   11 12 13 14 15 16 17    9 10 11 12 13 14 15   13 14 15 16 17 18 19
    21 22 23 24 25 26 27   18 19 20 21 22 23 24   16 17 18 19 20 21 22   20 21 22 23 24 25 26
    28 29 30 31            25 26 27 28 29 30      23 24 25 26 27 28 29   27 28 29 30 31      
                                                  30 31                                      
          September               October               November               December      
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
                    1  2    1  2  3  4  5  6  7             1  2  3  4                   1  2
     3  4  5  6  7  8  9    8  9 10 11 12 13 14    5  6  7  8  9 10 11    3  4  5  6  7  8  9
    10 11 12 13 14 15 16   15 16 17 18 19 20 21   12 13 14 15 16 17 18   10 11 12 13 14 15 16
    17 18 19 20 21 22 23   22 23 24 25 26 27 28   19 20 21 22 23 24 25   17 18 19 20 21 22 23
    24 25 26 27 28 29 30   29 30 31               26 27 28 29 30         24 25 26 27 28 29 30
                                                                         31                  

    cal("03")

             2018        

            March                                                                            
    Su Mo Tu We Th Fr Sa                                                                     
                 1  2  3                                                                     
     4  5  6  7  8  9 10                                                                     
    11 12 13 14 15 16 17                                                                     
    18 19 20 21 22 23 24                                                                     
    25 26 27 28 29 30 31                                                                     
                                                                                             

    cal("Jan")

             2018        

           January                                                                           
    Su Mo Tu We Th Fr Sa                                                                     
        1  2  3  4  5  6                                                                     
     7  8  9 10 11 12 13                                                                     
    14 15 16 17 18 19 20                                                                     
    21 22 23 24 25 26 27                                                                     
    28 29 30 31                                                                              
                                                                                             

    cal("december")

             2018        

          December                                                                           
    Su Mo Tu We Th Fr Sa                                                                     
                       1                                                                     
     2  3  4  5  6  7  8                                                                     
     9 10 11 12 13 14 15                                                                     
    16 17 18 19 20 21 22                                                                     
    23 24 25 26 27 28 29                                                                     
    30 31                                                                                    

    cal("2:5")

                                               2018                                           

          February                 March                  April                   May        
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
                 1  2  3                1  2  3    1  2  3  4  5  6  7          1  2  3  4  5
     4  5  6  7  8  9 10    4  5  6  7  8  9 10    8  9 10 11 12 13 14    6  7  8  9 10 11 12
    11 12 13 14 15 16 17   11 12 13 14 15 16 17   15 16 17 18 19 20 21   13 14 15 16 17 18 19
    18 19 20 21 22 23 24   18 19 20 21 22 23 24   22 23 24 25 26 27 28   20 21 22 23 24 25 26
    25 26 27 28            25 26 27 28 29 30 31   29 30                  27 28 29 30 31      
                                                                                             

    cal("-r 4")

                                               2018                                           

            March                  April                   May                   June        
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
                 1  2  3    1  2  3  4  5  6  7          1  2  3  4  5                   1  2
     4  5  6  7  8  9 10    8  9 10 11 12 13 14    6  7  8  9 10 11 12    3  4  5  6  7  8  9
    11 12 13 14 15 16 17   15 16 17 18 19 20 21   13 14 15 16 17 18 19   10 11 12 13 14 15 16
    18 19 20 21 22 23 24   22 23 24 25 26 27 28   20 21 22 23 24 25 26   17 18 19 20 21 22 23
    25 26 27 28 29 30 31   29 30                  27 28 29 30 31         24 25 26 27 28 29 30
                                                                                             
            July                                                                             
    Su Mo Tu We Th Fr Sa                                                                     
     1  2  3  4  5  6  7                                                                     
     8  9 10 11 12 13 14                                                                     
    15 16 17 18 19 20 21                                                                     
    22 23 24 25 26 27 28                                                                     
    29 30 31                                                                                 
                                                                                             

    cal("-r -3")


        December 2017          January 2018           February 2018           March 2018     
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
                    1  2       1  2  3  4  5  6                1  2  3                1  2  3
     3  4  5  6  7  8  9    7  8  9 10 11 12 13    4  5  6  7  8  9 10    4  5  6  7  8  9 10
    10 11 12 13 14 15 16   14 15 16 17 18 19 20   11 12 13 14 15 16 17   11 12 13 14 15 16 17
    17 18 19 20 21 22 23   21 22 23 24 25 26 27   18 19 20 21 22 23 24   18 19 20 21 22 23 24
    24 25 26 27 28 29 30   28 29 30 31            25 26 27 28            25 26 27 28 29 30 31
    31                                                                                       

    cal("-r -3:4")


        December 2017          January 2018           February 2018           March 2018     
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
                    1  2       1  2  3  4  5  6                1  2  3                1  2  3
     3  4  5  6  7  8  9    7  8  9 10 11 12 13    4  5  6  7  8  9 10    4  5  6  7  8  9 10
    10 11 12 13 14 15 16   14 15 16 17 18 19 20   11 12 13 14 15 16 17   11 12 13 14 15 16 17
    17 18 19 20 21 22 23   21 22 23 24 25 26 27   18 19 20 21 22 23 24   18 19 20 21 22 23 24
    24 25 26 27 28 29 30   28 29 30 31            25 26 27 28            25 26 27 28 29 30 31
    31                                                                                       
         April 2018              May 2018               June 2018              July 2018     
    Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa   Su Mo Tu We Th Fr Sa
     1  2  3  4  5  6  7          1  2  3  4  5                   1  2    1  2  3  4  5  6  7
     8  9 10 11 12 13 14    6  7  8  9 10 11 12    3  4  5  6  7  8  9    8  9 10 11 12 13 14
    15 16 17 18 19 20 21   13 14 15 16 17 18 19   10 11 12 13 14 15 16   15 16 17 18 19 20 21
    22 23 24 25 26 27 28   20 21 22 23 24 25 26   17 18 19 20 21 22 23   22 23 24 25 26 27 28
    29 30                  27 28 29 30 31         24 25 26 27 28 29 30   29 30 31            
                                                                                             

    cal("03 2015")

             2015        

            March                                                                            
    Su Mo Tu We Th Fr Sa                                                                     
     1  2  3  4  5  6  7                                                                     
     8  9 10 11 12 13 14                                                                     
    15 16 17 18 19 20 21                                                                     
    22 23 24 25 26 27 28                                                                     
    29 30 31                                                                                 
                                                                                             

    cal("Jan 2015")

             2015        

           January                                                                           
    Su Mo Tu We Th Fr Sa                                                                     
                 1  2  3                                                                     
     4  5  6  7  8  9 10                                                                     
    11 12 13 14 15 16 17                                                                     
    18 19 20 21 22 23 24                                                                     
    25 26 27 28 29 30 31                                                                     
                                                                                             

    cal("January 2015")

             2015        

           January                                                                           
    Su Mo Tu We Th Fr Sa                                                                     
                 1  2  3                                                                     
     4  5  6  7  8  9 10                                                                     
    11 12 13 14 15 16 17                                                                     
    18 19 20 21 22 23 24                                                                     
    25 26 27 28 29 30 31                                                                     
                                                                                             

Contact
=======

You are welcome to:  
- submit suggestions and bug-reports at:
<https://github.com/trinker/cal/issues>  
- send a pull request on: <https://github.com/trinker/cal>  
- compose a friendly e-mail to: <tyler.rinker@gmail.com>


NEWS
NEWS 
====

Versioning
----------

Releases will be numbered with the following semantic versioning format:

<major>.<minor>.<patch>

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor 
  and patch)
* New additions without breaking backward compatibility bumps the minor 
  (and resets the patch)
* Bug fixes and misc changes bumps the patch


cal 0.0.1 
----------------------------------------------------------------

This package is...


inst/maintenance.R
#========
# BUILD
#========
update_date(); update_news(); source("inst/build.R")

#==========================
# Run unit tests
#==========================
devtools::test()

#==========================
# knit README.md
#==========================
rmarkdown::render("README.Rmd", "all");# md_toc()#

#==========================
# UPDATE NEWS
#==========================
update_news()

#==========================
# UPDATE VERSION
#==========================
update_version()

#========================
#staticdocs dev version
#========================

if (!require("pacman")) install.packages("pacman")
pacman::p_load_gh("hadley/staticdocs", "trinker/acc.roxygen2")
p_load(rstudioapi, qdap)

R_USER <-  switch(Sys.info()[["user"]],
    Tyler = "C:/Users/Tyler",
    trinker = "C:/Users/trinker",
    message("Computer name not found")
)
build_site(pkg=file.path(R_USER, "GitHub", basename(getwd())), launch = FALSE)

#STEP 2: reshape index
path <- "inst/web"
path2 <- file.path(path, "/index.html")
rdme <- file.path(R_USER, "GitHub", basename(getwd()), "inst/extra_statdoc/readme.R")

extras <- qcv("")
## drops <- qcv()
expand_statdoc(path2, to.icon = extras, readme = rdme)

x <- readLines(path2)
x[grepl("<h2>Authors</h2>", x)] <- paste(
    c("<h2>Author</h2>"
    #rep("<h2>Contributor</h2>", 1)
    ),
    c("Tyler W. Rinker")
)

cat(paste(x, collapse="\n"), file=path2)

#STEP 3: move to trinker.guthub
library(reports)
file <- file.path(R_USER, "/GitHub/trinker.github.com")
# incoming <- file.path(file, basename(getwd()))
delete(incoming)
file.copy(path, file, TRUE, TRUE)
file.rename(file.path(file, "web"), incoming)
## delete(path)

#==========================
#staticdocs current version
#==========================

#STEP 3: move to trinker.guthub
library(reports)
file <- file.path(R_USER, "/GitHub/trinker.github.com")
incoming <- file.path(file, "discon")
##  delete(incoming); file.copy(path, file, TRUE, TRUE); file.rename(file.path(file, "web"), incoming)

#==========================
# NEWS new version
#==========================
x <- c("BUG FIXES", "NEW FEATURES", "MINOR FEATURES", "IMPROVEMENTS", "CHANGES")
cat(paste(x, collapse = "\n\n"), file="clipboard")





R/cal.R
#' Command Line Calendar
#'
#' A command line calendar in the style of the \pkg{Unix} \pkg{cal} command
#' line tool though with only only flag.
#'
#' @param cmd A string in the form of Month (int/char) AND/OR Year (4 digits).
#' Below are the forms that are accepted:
#' \describe{
#'   \item{2017}{A full calendar for the year}
#'   \item{03}{A single month view}
#'   \item{Jan}{A single month view (case insensitive)}
#'   \item{January}{A single month view (case insensitive)}
#'   \item{2:5}{A multi month view (integers must be > 0 and < 13)}
#'   \item{-r 4}{A multi month view relative to the current month (curent month + n more; integers must be positive and < 13)}
#'   \item{-r -3:4}{A multi month view relative to the current month (curent month +/- n before/after; -r tag not required; lower must be negative; must be less than or equal to 12 months all together)}
#'   \item{03 2015}{A single month view for a particular year}
#'   \item{Jan 2015}{A single month view for a particular year (case insensitive)}
#'   \item{January 2015}{A single month view for a particular year (case insensitive)}
#' }
#' The default is \code{'-r 2'} which prints this month and the next 2.
#' @param ncols The number of columns to form the print grid.  This can be set
#' globally via \code{options(cal.ncols = 4)}
#' @param \ldots ignored.
#' @return Prints calendar to the command line.
#' @references \url{https://www.tutorialspoint.com/unix_commands/cal.htm}
#' @export
#' @examples
#' cal()
#' cal("2017")
#' cal("03")
#' cal("Jan")
#' cal("december")
#' cal("2:5")
#' cal("-r 4")
#' cal("-r -3")
#' cal("-r -3:4")
#' cal("03 2015")
#' cal("Jan 2015")
#' cal("January 2015")
cal <- function(cmd = '-r 2', ncols = NULL, ...){

    if (cmd == '0') cmd <- '-r 0'

    year_month <- parse_cmd(cmd)
    include.year <- length(unique(year_month$year)) == 1

    cells <- stats::setNames(
        Map(
            function(a, b){get_days(a, b, include.year = !include.year)},
            year_month$month,
            year_month$year
        ),
        year_month$year
    )

    if (length(cells) == 0) stop('Could not parse the `cmd`.  See `?cal` for proper formatting', call. = FALSE)

    out <- list(
        cells = cells,
        year_header = include.year,
        year = ifelse(include.year, year_month$year[1], NA)
    )

    class(out) <- 'cal'
    attributes(out)[['ncols']] <- ncols
    out
}






#' Prints a cal Object
#'
#' Prints a cal object
#'
#' @param x A cal Object
#' @param ncols The number of columns to form the print grid
#' @param \ldots ignored
#' @method print cal
#' @export
print.cal <- function(x, ncols = getOption("pax.ncols"), ...){

    if (is.null(ncols) && !is.null(attributes(x)[['ncols']])) {
        ncols <- attributes(x)[['ncols']]
    }

    if (is.null(ncols)) {
        width <- getOption('width')
        ncols <- 1 + floor((width - 20)/23)
    }

    len <- length(x[['cells']])
    nrows <- ceiling(len/ncols)

    fills <- lapply(seq_len((ncols * nrows) - len), function(i) blank)

    cells <- unlist(list(x[['cells']], fills), recursive = FALSE)

    if(x[['year_header']]) {

        header <- x[['year']]

        calwidth <- (min(ncols, len) * 23) - 2
        space <- calwidth - nchar(header)

        header <- paste(c(paste(c(
            paste(rep(' ', ceiling((calwidth - nchar(header))/2)), collapse = ''),
            header,
            paste(rep(' ', floor((calwidth - nchar(header))/2)), collapse = '')
        ), collapse = ''), ''), collapse = '\n')

    } else {
        header <- NULL
    }

    out <- paste(unlist(lapply(
        split(cells, rep(seq_len(nrows), each = ncols)),
        function(x){
            paste(unlist(apply(data.frame(lapply(x, c)), 1, paste, collapse = '   ')), collapse ='\n')

        }
    )), collapse = '\n')

    cat(header, out, sep = '\n')
}

## A constant for adding blank cells
blank <- rep(paste(rep(' ', 20), collapse = ''), 8)



# test_str <- c(
#     ## entire year
#     '2018',
#
#     ## multiple months (fail this later if > 12 output)
#     '-r 3',
#     '-r -3',
#     '-r -3:4',
#     '-2:4',
#     '4:5',
#
#     ## single months
#     '01',
#     '1',
#     'Jan',
#     'feb',
#     'March',
#     'december',
#
#     ## month year
#     '01 2018',
#     '1 2018',
#     'Jan 2018',
#     'feb 2018',
#     'March 2018',
#     'december 2018',
#
#     NA,
#     'sd ds',
#     '123',
#     '0',
#     'dece'
# )




## regex checks
##=============
r_relative_num <- '-r\\s+-*\\d{1,2}'
r_relative_vect <- '(-r\\s+)?-\\d{1,2}:\\d{1,2}'

## convert first 2 with look up table
r_month_abb <- paste0('(', paste(tolower(month.abb), collapse = '|'), ')')
r_month_name <- paste0('(', paste(tolower(month.name), collapse = '|'), ')')
r_month_num <- '(0?[1-9]|1[012])'
r_months_num <- '\\d{1,2}:\\d{1,2}'

## strip r eval
r_year <- '\\d{4}'


regex_forms <- stats::setNames(
    as.list(paste0('^(', c(
        paste(c(r_month_abb, r_month_name, r_month_num), r_year),
        r_relative_num,
        r_relative_vect,
        r_month_abb,
        r_month_name,
        r_month_num,
        r_months_num,
        r_year
    ), ')$')),
    c(
        paste(c('r_month_abb', 'r_month_name', 'r_month_num'), 'year', sep ='_'),
        "r_relative_num", "r_relative_vect", "r_month_abb", "r_month_name", "r_month_num", "r_months_num", "r_year"
    )
)


## function to parse the cmd string
parse_cmd <- function(cmd, ...){

    cmd <- tolower(cmd)
    type <- get_cmd_type(cmd)

    cmonth <- as.integer(format(Sys.Date(), format = '%m'))
    cyear <- as.integer(format(Sys.Date(), format = '%Y'))

    ## return = list(months, years), parse m/y add m/y where missing
    switch(type,

        r_month_abb_year = {

            list(
                year = as.integer(gsub('(^[^ ]+\\s+)(\\d{4}$)', '\\2', cmd)),
                month = match(gsub('(\\s+\\d{4}$)', '', cmd), tolower(month.abb))
            )

        },

        r_month_name_year = {

            list(
                year = as.integer(gsub('(^[^ ]+\\s+)(\\d{4}$)', '\\2', cmd)),
                month = match(gsub('(\\s+\\d{4}$)', '', cmd), tolower(month.name))
            )

        },

        r_month_num_year = {

            list(
                year = as.integer(gsub('(^\\d{1,2}\\s+)(\\d{4}$)', '\\2', cmd)),
                month = as.integer(gsub('(^\\d{1,2})(\\s+\\d{4}$)', '\\1', cmd))
            )

        },

        r_relative_num = {

            months <- 0:eval(parse(text = gsub('-r\\s+', '', cmd)))
            months2 <- sort(cmonth + months)

            list(
                year = c(
                    rep(cyear - 1, length(months2[months2 <= 0])),
                    rep(cyear, length(months2[months2 > 0 & months2 < 13])),
                    rep(cyear + 1, length(months2[months2 > 12]))
                ),
                month = c(
                    12 - abs(months2[months2 <= 0]),
                    months2[months2 > 0 & months2 < 13],
                    months2[months2 > 12] - 12
                )
            )

        },

        r_relative_vect = {

            months <- eval(parse(text = gsub('-r\\s+', '', cmd)))
            if (length(months) > 12) stop('Number of months must be less or equal to 12.', call. = FALSE)

            months2 <- cmonth + months

            list(
                year = c(
                    rep(cyear - 1, length(months2[months2 <= 0])),
                    rep(cyear, length(months2[months2 > 0 & months2 < 13])),
                    rep(cyear + 1, length(months2[months2 > 12]))
                ),
                month = c(
                    12 - abs(months2[months2 <= 0]),
                    months2[months2 > 0 & months2 < 13],
                    months2[months2 > 12] - 12
                )
            )

        },

        r_month_abb = {

            months <- match(cmd, tolower(month.abb))

            list(
                year = rep(cyear, length(months)),
                month = months
            )

        },

        r_month_name = {

            months <- match(cmd, tolower(month.name))

            list(
                year = rep(cyear, length(months)),
                month = months
            )

        },

        r_month_num = {

            list(
                year = cyear,
                month = as.integer(cmd)
            )

        },

        r_months_num = {

            months <- eval(parse(text = cmd))
            if (length(months) > 12) stop('Number of months must be less or equal to 12.', call. = FALSE)
            if (max(months) > 12 | min(months) < 1) stop('Month itnegers cannot be > 12 or < 1', call. = FALSE)

            list(
                year = rep(cyear, length(months)),
                month = months
            )

        },

        r_year = {

            list(
                year = rep(as.integer(cmd), 12),
                month = 1:12
            )

        },

        stop('`cmd` was not a valid format.  Use `?cal` to see valid formats for `cmd`.', call. = FALSE)
    )

}


## short cut for grepl to validate strings (detect if a regex cond is matched)
validator <- function(str, reg) grepl(reg, str, ignore.case = TRUE)

## function to determine which of the cal formats that cmd is
get_cmd_type <- function(cmd, regs = regex_forms, ...){

    unlist(lapply(cmd, function(x){

        out <- unlist(lapply(regs, function(y) validator(x, y)))
        out <- names(out)[out]
        if (length(out) == 0) out <- NA
        out

    }))
}



## function to grab the days of a month and format them for that year
get_days <- function(month = NULL, year = NULL, include.year = FALSE, ...){

    if (is.null(month)) {
        month <- as.integer(format(Sys.Date(), format = '%m'))
    }
    if (is.null(year)) {
        year <- as.integer(format(Sys.Date(), format = '%Y'))
    }

    ## number of days and start day of week
    start_date <- as.Date(sprintf('%s-%s-01', year, month))
    days <- stringi::stri_pad(
        seq_len(lubridate::days_in_month(start_date)),
        width = 2,
        side = 'left',
        pad = ' '
    )

    start_dow <- weekdays(start_date)

    front_fills <- rep('  ', lubridate::wday(start_date) - 1)

    days_filled <- c(front_fills, days)
    days_filled <- c(days_filled, rep('  ', 42 - length(days_filled)))

    out <- apply(matrix(days_filled, ncol = 7, byrow = TRUE), 1, paste, collapse = ' ')
    month <- format(start_date, format = '%B')

    if (include.year) month <- paste(month, year)

    out <- c(
        paste(c(
            paste(rep(' ', ceiling((nchar(out[1]) - nchar(month))/2)), collapse = ''),
            month,
            paste(rep(' ', floor((nchar(out[1]) - nchar(month))/2)), collapse = '')
        ), collapse = ''),
        paste(c('Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'), collapse = ' '),
        out
    )

    class(out) <- 'calmonth'
    out
}

